cmd/eventkitd-sink-bigquery/main.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"net"
+	"os"
+	"os/signal"
+	"runtime"
+	"syscall"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"github.com/jtolio/eventkit/pb"
+	"github.com/jtolio/eventkit/transport"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Application struct {
+	Name     string
+	Instance string
+	Version  string
+}
+
+type Record struct {
+	Application Application
+
+	StartTime  time.Time
+	ReceivedAt time.Time
+	SendOffset time.Duration
+
+	Source *net.UDPAddr
+
+	Tags []*pb.Tag
+}
+
+func (r *Record) Save() (map[string]bigquery.Value, string, error) {
+	fields := make(map[string]bigquery.Value)
+	fields["application_name"] = r.Application.Name
+	fields["application_instance"] = r.Application.Instance
+	fields["application_version"] = r.Application.Version
+	fields["start_time"] = r.StartTime
+	fields["received_at"] = r.ReceivedAt
+	fields["send_offset"] = r.SendOffset
+	fields["source_ip"] = r.Source.IP.String()
+
+	for _, tag := range r.Tags {
+		field := "tag_" + tag.Key
+
+		switch v := tag.Value.(type) {
+		case *pb.Tag_Bool:
+			fields[field] = v.Bool
+		case *pb.Tag_Bytes:
+			fields[field] = v.Bytes
+		case *pb.Tag_Double:
+			fields[field] = v.Double
+		case *pb.Tag_Duration:
+			fields[field] = v.Duration
+		case *pb.Tag_Int64:
+			fields[field] = v.Int64
+		case *pb.Tag_String_:
+			fields[field] = v.String_
+		}
+	}
+
+	return fields, "", nil
+}
+
+// BigQuerySink provides an abstraction for processing events in a transport agnostic way.
+type BigQuerySink struct {
+	dataset *bigquery.Dataset
+}
+
+// Receive is called when the server receive an event to process.
+func (b *BigQuerySink) Receive(ctx context.Context, packet *Packet) error {
+	records := make(map[string][]*Record)
+
+	for _, event := range packet.Packet.Events {
+		for _, scope := range event.Scope {
+			k := scope + "." + event.Name
+
+			records[k] = append(records[k], &Record{
+				Application: Application{
+					Name:     packet.Packet.Application,
+					Instance: packet.Packet.Instance,
+					Version:  packet.Packet.ApplicationVersion,
+				},
+				StartTime:  packet.Packet.StartTimestamp.AsTime(),
+				ReceivedAt: packet.ReceivedAt,
+				SendOffset: packet.Packet.SendOffset.AsDuration(),
```

Comment:
perhaps we can reuse the record code that already exists in eventkitd?

Reply:
I'm open to that (or at least having some common record code somewhere)... the only reason I pulled the bigquery stuff out was so the core package didn't have a ton of Google dependendencies.